---
title: "edgeR"
author: "Dennis Amnebrink"
date: "7/27/2022"
output:
  html_document:
      toc: yes
      toc_float:
        collapse: no
      fig_caption: yes
      code_folding: hide
---
# Comparing responses to nutrient inputs at different temperature regimes
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(edgeR)
library(data.table)
library(viridis)
library(kableExtra)
```

```{r file-loading, message=FALSE, warning=FALSE, cache=TRUE}
#File to connect NGI ID with our own classification ID, add the Âµ-values to this file for testing later, or do that in a sep file?
sample_ID <- read_tsv("../data/J.Pinhassi_21_02_sample_info.txt") %>%
  dplyr::select(1:2) %>%
  dplyr::rename(sample = 1, sample_name = 2) %>%
  mutate(sample_name = str_replace(sample_name, "[0-9]*:","")) %>%
  separate(sample_name, c("treatment","timepoint"), sep = ",") %>%
  mutate(treatment = gsub("NT","TN", treatment)) %>%
  mutate(sample_name = paste(timepoint, treatment, sep = "_")) %>%
  separate(treatment, c("treatment","replicate"), sep = -1) %>%
  mutate(tre_rep = gsub("_","", sample_name)) %>%
  mutate(tre_rep = toupper(tre_rep)) %>%
  mutate(day = case_when( timepoint == "t3" ~ 10,
                          TRUE ~ 17
                          )
         )

#Reading in annotations
eggnogs <- read_tsv("../data/eggnog_annotations.tsv.gz")

#Reading in taxonomic annotation of reads
taxonomy <- read_tsv("../data/eukulele_phylodb.tsv") %>%
  select(-species) %>%
  rename(species = "genus", genus = "family", family = "order", order = "class", class = "phylum", phylum = "kingdom")
  
#Count file for all orfs, tpm is calculated per sample
bbmap <- fread("../data/bbmap_counts.tsv.gz", sep = "\t") %>%
              filter(count > 0)  %>%                               # removing 0 counts to reduce size of table
              mutate(Geneid = str_replace(Geneid, "[0-9]*_",""),   # Joining in chr and geneid to match format of eggnog naming.
                     orf = paste(Chr,Geneid, sep = "_"),
                     sample = str_extract(sample, "P[0-9]*_[0-9]*")) %>% # Removing extra numbers in sample to match sample_ID format
              dplyr::select(-Geneid,-Chr) %>%
              dplyr::select(orf,Start,End,Strand,Length,sample,count,tpm) 
```

```{r constants, message=FALSE, warning=FALSE}
SIGNIFICANCE = 0.05
DIFF_TRESHOLD = 2.5
```

```{r R_colours}
GEOM_COL_COLOURS = c(
  RColorBrewer::brewer.pal(12, name = 'Paired'))
```

```{r prok-filter, message=FALSE, warning=FALSE, cache=TRUE}
#Removing non-prokreads as well as recalculating tpm over prok frac.
bbmap_p <- bbmap %>% 
  semi_join(taxonomy %>% filter(domain %in% c("Bacteria", "Archaea")), 
            by = "orf") %>%
  dplyr::select(-tpm) %>%
   group_by(sample) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
   ungroup()
```

# EdgeR-analysis KO-terms day 10 
```{r edgeR_setup, message=FALSE, warning=FALSE}
# Do N vs C only day 10 (T3)
 bbmap_mat <- bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  inner_join(eggnogs %>% distinct(orf, KEGG_ko), by = "orf") %>%
  filter(KEGG_ko != "-") %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  group_by(sample, KEGG_ko) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "10") %>% #Keeping day 10
  select(sample, KEGG_ko, count) %>%
  spread(sample, count, fill = 0) %>%
  column_to_rownames("KEGG_ko") %>%
  as.matrix()
 
library(edgeR)

y <- DGEList(bbmap_mat)

#Changing the grouping to the known treatments from mesocosm.
y$samples$group = sample_ID %>% filter(day == "10") %>% select(sample, treatment) %>%  pull(treatment)

#It has been correctly done as they match with the original sample file
groupF <- y$samples$group

#Another way of verifying
match(rownames(y$samples), colnames(bbmap_mat))

y <- DGEList(counts=bbmap_mat, group=groupF)

keep <- filterByExpr(y, group=groupF)

y <- y[keep, , keep.lib.sizes=FALSE]

y <- calcNormFactors(y)

y$samples

#The design-matrix doesn't like non-numeric categories, so here is a numeric corresponding to orignal levels, key =
# 1 = T, 2 = C, 3 = N, 4 = TN
design_m <- y$samples %>% select(group) %>%
  mutate(group = c(1,1,1,2,2,2,3,3,3,4,4,4))

y <- estimateDisp(y, design = design_m)

plotBCV(y)

#Creating a designmatrix (this looks like the correct structure), C is the base treatment, exlcuding intercept column.
trial_mat <- model.matrix(~0+groupF)

fit <- glmQLFit(y, trial_mat)

fit

#Comparing C vs T
colnames(trial_mat) = c("C","N","TN","t")

fit <- glmQLFit(y, trial_mat)

TvsC <- makeContrasts(t - C, levels=trial_mat)
NvsC <- makeContrasts(N - C, levels=trial_mat)
TNvsT <- makeContrasts(TN - t, levels=trial_mat)
```

```{r TvsC, message=FALSE, warning=FALSE}
qlf <- glmQLFTest(fit, contrast=TvsC)

#Most changing genes
topTags(qlf)

#Summary of genes up/downregulated, upregulated in this case refes to upregulated in C-treatment, negative is upregulated in TN-treatment.

summary(decideTests(qlf))
# Blue lines indicate 2-fold changes

pdf("../results/edgeRTvsC2logFC.pdf")
plotMD(qlf)
abline(h=c(-1, 1), col="blue")
dev.off()

#Trying to look at the fold changes the edgeR way
tr <- glmTreat(fit, contrast=TvsC, lfc=log2(DIFF_TRESHOLD))

#The summary lets us know there are 1015 genes upregulated in  the C-treatment, and 4045 upregulated in  the TN treatment at a 2.5 logFC value.
diff_expr_sum <- summary(decideTestsDGE(tr)) %>% as.data.frame()

write_tsv(diff_expr_sum, "../results/edgeR_gene_partitioningTvsC.tsv")

TvsC <- tr$table %>%
  rownames_to_column("KEGG_ko")

#Doing the same but setting treshold to 2.5 logFC
pdf("../results/edgeRTvsC2logFC2.5.pdf")
plotMD(tr)
abline(h=c(-DIFF_TRESHOLD, DIFF_TRESHOLD), col="blue")
dev.off()
```

```{r NvsC, message=FALSE, warning=FALSE}
qlf <- glmQLFTest(fit, contrast=NvsC)

#Most changing genes
topTags(qlf)

#Summaroy of genes up/downregulated, upregulated in this case refes to upregulated in C-treatment, negative is upregulated in TN-treatment.

summary(decideTests(qlf))
# Blue lines indicate 2-fold changes

pdf("../results/edgeRNvsC2logFC.pdf")
plotMD(qlf)
abline(h=c(-1, 1), col="blue")
dev.off()

#Trying to look at the fold changes the edgeR way
tr <- glmTreat(fit, contrast=NvsC, lfc=log2(DIFF_TRESHOLD))

NvsC <- tr$table %>%
  rownames_to_column("KEGG_ko")
```

```{r TNvsT, message=FALSE, warning=FALSE}
qlf <- glmQLFTest(fit, contrast=TNvsT)

#Most changing genes
topTags(qlf)

#Summaroy of genes up/downregulated, upregulated in this case refes to upregulated in C-treatment, negative is upregulated in TN-treatment.

summary(decideTests(qlf))
# Blue lines indicate 2-fold changes

pdf("../results/edgeRTNvsT2logFC.pdf")
plotMD(qlf)
abline(h=c(-1, 1), col="blue")
dev.off()

#Trying to look at the fold changes the edgeR way
tr <- glmTreat(fit, contrast=TNvsT, lfc=log2(DIFF_TRESHOLD))

TNvsT <- tr$table %>%
  rownames_to_column("KEGG_ko")
```

```{r message=FALSE, warning=FALSE}
#Combiningputput df's and performing fdr correction
TvsC %>%
  as.data.frame() %>%
  mutate(contrast = "TvsC") %>%
  bind_rows(NvsC %>% mutate(contrast = "NvsC")) %>%
  bind_rows(TNvsT %>% mutate(contrast = "TNvsT")) %>%
  mutate(fdr = p.adjust(PValue, method = "fdr")) %>%
  mutate(type = 
           case_when(
             (fdr <= SIGNIFICANCE & logFC >= DIFF_TRESHOLD) ~ "Up",
             (fdr <= SIGNIFICANCE & logFC <= -DIFF_TRESHOLD) ~ "Down",
             TRUE ~"non-DE"
           )) %>%
  mutate(type = forcats::as_factor(type) %>% fct_relevel("Up","non-DE","Down")) %>%
  write_tsv("../results/edgeR_contrasts.tsv")
```

## Inspecting contrasts
```{r TvsC_KO, message=FALSE, warning=FALSE, results=FALSE}
contrasts <- read_tsv("../results/edgeR_contrasts.tsv")

mat <- contrasts %>%
  filter(contrast == "TvsC") %>%
  filter(fdr <= SIGNIFICANCE) %>%
  arrange(desc(logFC)) %>%
  mutate(KEGG_ko = factor(KEGG_ko) %>% forcats::fct_reorder(logFC, .desc = TRUE)) %>%
  select(KEGG_ko, logFC) %>%
  spread(KEGG_ko, logFC, fill = 0) %>%
  as.matrix() 
 
 
 #use geom rug to add the colourstrip on the X axis
bbmap_plot <- bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  inner_join(eggnogs %>% distinct(orf, KEGG_ko), by = "orf") %>%
  filter(KEGG_ko != "-") %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  group_by(sample, KEGG_ko) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "10") %>%#Removing day 10 
  filter(treatment %in% c("C","T")) %>%
  group_by(treatment, KEGG_ko) %>%
  summarise(mean = mean(count), stdev = sd(count)) %>%
  ungroup() %>%
  inner_join(contrasts %>%
    filter(contrast == "TvsC") %>%
    filter(fdr <= SIGNIFICANCE) %>%
    arrange(desc(logFC)) %>%
    select(KEGG_ko, logFC), by = "KEGG_ko") %>%
  mutate(KEGG_ko = factor(KEGG_ko) %>% forcats::fct_reorder(logFC, .desc = TRUE)) 
  
bbmap_plot %>% 
  filter(mean <= 80000) %>%
  #filter(logFC>= 4 | logFC <= -4) %>%
 ggplot(mapping = aes(x = KEGG_ko, y = log(mean))) +
  geom_col( aes(fill = treatment)) +
  geom_rug(aes(colour = logFC), sides = "b") +
  scale_colour_viridis(discrete = FALSE) +
  theme(axis.text.x = element_blank()) +
  facet_wrap(~treatment)
  ggsave("../results/TvsC_KO_logFC.png")
  
  
contrasts %>% 
  filter(contrast == "TvsC") %>%
  filter(fdr <= 0.05) %>%
 ggplot(mapping = aes(x = KEGG_ko, y = logFC, colour=logCPM)) +
  geom_point(aes(size= fdr)) +
  scale_colour_viridis(discrete = FALSE) +
  theme(axis.text.y = element_blank(), axis.ticks.y =element_blank()) +
  coord_flip()
```
In the TvsC contrast there were `r nrow(contrasts %>% filter(contrast == "TvsC") %>% filter(fdr <= 0.05 & logFC >= DIFF_TRESHOLD))` genes  significantly upregulated. And `r nrow(contrasts %>% filter(contrast == "TvsC") %>% filter(fdr <= 0.05 & logFC <= -DIFF_TRESHOLD))` Downregulated. Out of a total of `r nrow(contrasts %>% filter(contrast =="TvsC"))`genes.
## Alternative visualization of the TvsC contrast significant genes.
```{r TvsC_KO_plot, message=FALSE, warning=FALSE, fig.cap = "Differentially expressed genes C vs T"}

DE_genes <- contrasts %>%
   filter(contrast == "TvsC") %>%
   filter(type %in% c("Up","Down"))

bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  inner_join(eggnogs %>% distinct(orf, KEGG_ko), by = "orf") %>%
  filter(KEGG_ko != "-") %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  group_by(sample, KEGG_ko) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "10") %>% #Removing day 10
  select(sample, KEGG_ko, count, treatment) %>%
  filter(treatment %in% c("C","T")) %>%
  mutate(KEGG_ko = gsub("ko:","", KEGG_ko)) %>%
  semi_join(DE_genes %>% filter(type == "Down"), by = "KEGG_ko") %>%
  group_by(treatment, KEGG_ko) %>%
  summarise(count= mean(count)) %>%
  ungroup() %>%
  ggplot(mapping = aes(x = KEGG_ko, y = count, fill = treatment)) +
  geom_col()
```
## Overview stats

This table shows the number of significantly upregulated/downregulated KO-terms as well as non-significant KO-terms per contrast. 
```{r}
kable(contrasts %>%
  group_by(contrast, type) %>%
  tally() %>%
  ungroup() %>%
  spread(contrast, n, fill = 0)) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
  add_header_above(c("Table 1. EdgeR stats contrasts" = 4), font_size = 18) %>%
  scroll_box(width = "100%", height = "200px")
```

## NvsC
In the NvsC contrast there were `r nrow(contrasts %>% filter(contrast == "NvsC") %>% filter(fdr <= 0.05 & logFC >= DIFF_TRESHOLD))` genes  significantly upregulated. And `r nrow(contrasts %>% filter(contrast == "NvsC") %>% filter(fdr <= 0.05 & logFC <= -DIFF_TRESHOLD))` Downregulated. Out of a total of `r nrow(contrasts %>% filter(contrast == "NvsC"))`KO-terms.
```{r NvsC_KO, message=FALSE, warning=FALSE, eval=FALSE}
DE_genes <- contrasts %>%
   filter(contrast == "NvsC") %>%
   filter(type %in% c("Up","Down"))

contrasts %>% 
  filter(contrast == "NvsC") %>%
  filter(fdr <= 0.05) %>%
 ggplot(mapping = aes(x = KEGG_ko, y = logFC, colour=logCPM)) +
  geom_point(aes(size= fdr)) +
  scale_colour_viridis(discrete = FALSE) +
  theme(axis.text.y = element_blank(), axis.ticks.y =element_blank()) +
  coord_flip()
```

```{r NvsC_volcano}
contrasts %>% 
  filter(contrast == "NvsC") %>%
  ggplot(mapping = aes(x=logCPM, y = logFC, colour=type, size = type)) +
  geom_point()  +
  scale_size_manual(values = c("Down" = 1.5, "Up"=1.5, "non-DE" = 0.3)) +
  scale_colour_manual(values = c("Down" = "Blue","Up" = "Red", "non-DE" = "Black")) +
  theme_minimal()
```
KO-term annotations were retrieved from the [KEGG-website](https://www.genome.jp/kegg/ko.html) By pasting the produced vector of KO-terms and copying the output into a new file.
```{r KO-term_identification, message=FALSE, warning=FALSE, results=FALSE}
contrasts %>%
filter(contrast == "NvsC" & type == "Up") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  mutate(gene = paste0("gene", row_number())) %>%
  select(gene, KEGG_ko) %>%
  write_tsv("../results/NvsC_Up.tsv")

contrasts %>%
filter(contrast == "NvsC" & type == "Down") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  mutate(gene = paste0("gene", row_number())) %>%
  select(gene, KEGG_ko) %>%
  write_tsv("../results/NvsC_Down.tsv")

contrasts %>%
filter(contrast == "NvsC" & type == "Up") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  mutate(gene = paste0("gene", row_number())) %>%
  select(gene, KEGG_ko) %>%
  pull(KEGG_ko)

contrasts %>%
filter(contrast == "NvsC" & type == "Down") %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  mutate(gene = paste0("gene", row_number())) %>%
  select(gene, KEGG_ko) %>%
  pull(KEGG_ko)
```

```{r}
NvsCUp <- read_tsv("../results/CvsN_up_legend.tsv", col_names = "KEGG_ko") %>%
  separate(KEGG_ko, c("KEGG_ko","gene"), sep = 9) %>%
  separate(gene, c ("gene","description"), sep = ";") %>%
  mutate(gene = gsub(" ","", gene))

kable(NvsCUp) %>%
  kable_styling(bootstrap_options = c("striped", "hover")) %>%
    add_header_above(c("Table. Upregulated KO's in NvsC" = 3), font_size = 18) %>%
  scroll_box(width = "100%", height = "200px")
```
Interesting genes included   
[aer](https://www.uniprot.org/uniprotkb/Q9I3F6/entry) - ionvolved in chemotaxis (aerotaxis).   
[arcA](https://www.uniprot.org/uniprotkb/P0A9Q1/entry) - [Two component system](https://www.mechanobio.info/pathogenesis/what-are-two-component-regulatory-systems/#:~:text=Two%2Dcomponent%20regulatory%20systems%20(TCRS,modulation%20of%20gene%20expression%20occurs.)   
[barA,gacS,varS](https://www.genome.jp/entry/K07678) - two component system.   
[cheA](https://www.uniprot.org/uniprotkb/P07363/entry) - two component system involved in chemotaxis     
[cheB](https://www.uniprot.org/uniprotkb/P04042/entry) - involved in chemotaxis   
[cheBR](https://www.uniprot.org/uniprotkb/A0A0P7Z799/entry) - involved in chemotaxis   
[cheV](https://www.uniprot.org/uniprotkb/P37599/entry) - invoved in chemotaxis   
[cheW](https://www.uniprot.org/uniprotkb/P0A964/entry) - involved in chemotaxis   
[cheZ](https://www.uniprot.org/uniprotkb/P0A9H9/entry) - involved in chemotaxis 
[cikA](https://www.uniprot.org/uniprotkb/A0A2K8WMD5/entry) - two component system   
[cphA](https://en.wikipedia.org/wiki/Cyanophycin) - polymer used during phosphate or sulfur starvation. 
[csGE](https://www.genome.jp/dbget-bin/www_bget?ko:K04337) - [curli production](https://en.wikipedia.org/wiki/Curli)   

[cydC](https://www.uniprot.org/uniprotkb/P23886/entry) - involved in transport   
[dcuB](https://www.uniprot.org/uniprotkb/P0ABN9/entry) - periplasm across inner membrane transport (C4).  
[FAB2, SSI2, desA1](https://www.genome.jp/dbget-bin/www_bget?ko:K03921) - Fatty acid biosynthesis.   
[fadB](https://www.uniprot.org/uniprotkb/P21177/entry) - Fatty acid beta oxidation   
[fadE](https://www.uniprot.org/uniprotkb/Q47146/entry) - Fatty acid beta oxidation   
[fadJ](https://www.uniprot.org/uniprotkb/P77399/entry) - FAtty acid beta oxidation   
[fadR](https://www.uniprot.org/uniprotkb/P0A8V6/entry) - Multifunctional regulator, represses genes required for fatty acid transport and beta-oxidation including fadA, fadB, fadD, fadL and fadE. Activates transcription of at least three genes required for unsaturated fatty acid biosynthesis: fabA, fabB and iclR, the gene encoding the transcriptional regulator of the aceBAK operon encoding the glyoxylate shunt enzymes.   
[gspC](https://www.uniprot.org/uniprotkb/P45757/entry) - Type II secretion, potentially involved in [signalling](https://en.wikipedia.org/wiki/Type_II_secretion_system).  
[K17213](https://www.genome.jp/dbget-bin/www_bget?ko+K17213) - ABC-transporter
[K17214](https://www.genome.jp/dbget-bin/www_bget?ko+K17214) - ABC-transporter
[lapA](https://www.genome.jp/dbget-bin/www_bget?ko+K12549) - surface adhesion protein in biofilm formation   
[luxO](https://www.genome.jp/dbget-bin/www_bget?ko+K10912) - two component system invlved in quorom sensing and biofilm formation.   
[luxS](https://www.genome.jp/dbget-bin/www_bget?ko+K07173) - involved in cystein methionine met, quorom sensing and biofilm formation.
[mcp](https://www.genome.jp/dbget-bin/www_bget?ko+K03406) - methyl accepting chemotaxis protein (two component system)   
[motY](https://www.genome.jp/dbget-bin/www_bget?ko+K21218) - flagellar assembly.   
[mscL](https://www.genome.jp/dbget-bin/www_bget?ko+K03282) - signalling and transporters.   
[mscS](https://www.genome.jp/dbget-bin/www_bget?ko+K03442) - signalling and transporters.     
[mshA](https://www.genome.jp/dbget-bin/www_bget?ko+K10924) - biofilm formation and T2SS.   
[mshB](https://www.genome.jp/dbget-bin/www_bget?ko+K10925) - biofilm formation and T2SS.   
[mshD](https://www.genome.jp/dbget-bin/www_bget?ko+K10927) - biofilm formation and T2SS.   
[mshI](https://www.genome.jp/dbget-bin/www_bget?ko+K12279) - MSHA biogenesis protein T2SS.   
[mshL](https://www.genome.jp/dbget-bin/www_bget?ko+K12282) - MSHA biogenesis protein T2SS.   
[mshP](https://www.genome.jp/dbget-bin/www_bget?ko+K12286) - MSHA biogenesis protein T2SS.   
[napA](https://www.genome.jp/dbget-bin/www_bget?ko+K02567) - nitrate reductase , Denitrification, nitrate => nitrogen, Dissimilatory nitrate reduction, nitrate => ammonia   
[napE](https://www.genome.jp/dbget-bin/www_bget?ko+K02571) - periplasmic nitrate reductase   
[narB](https://www.genome.jp/dbget-bin/www_bget?ko+K00367) - Nitrogen metabolism.




## TNvsT
In the NvsC contrast there were `r nrow(contrasts %>% filter(contrast == "TNvsT") %>% filter(fdr <= 0.05 & logFC >= DIFF_TRESHOLD))` genes  significantly upregulated. And `r nrow(contrasts %>% filter(contrast == "TNvsT") %>% filter(fdr <= 0.05 & logFC <= -DIFF_TRESHOLD))` Downregulated. Out of a total of `r nrow(contrasts %>% filter(contrast == "TNvsT"))`genes.
```{r TNvsT_KO, message=FALSE, warning=FALSE, eval=FALSE}
DE_genes <- contrasts %>%
   filter(contrast == "TNvsT") %>%
   filter(type %in% c("Up","Down"))

contrasts %>% 
  filter(contrast == "TNvsT") %>%
  filter(fdr <= 0.05) %>%
 ggplot(mapping = aes(x = KEGG_ko, y = logFC, colour=logCPM)) +
  geom_point(aes(size= fdr)) +
  scale_colour_viridis(discrete = FALSE) +
  theme(axis.text.y = element_blank(), axis.ticks.y =element_blank()) +
  coord_flip()
```

```{r TNvsT_volcano}
contrasts %>% 
  filter(contrast %in% c("TNvsT", "NvsC")) %>%
  ggplot(mapping = aes(x=logCPM, y = logFC, colour=type, size = type)) +
  geom_point()  +
  scale_size_manual(values = c("Down" = 1.5, "Up"=1.5, "non-DE" = 0.3)) +
  scale_colour_manual(values = c("Down" = "Blue","Up" = "Red", "non-DE" = "Black")) +
  theme_minimal() +
  facet_wrap(~ contrast)

ggsave("../results/nutrient_add_contrasts.pdf")
```

## Creating a key with pathway and KO-term
```{r message=FALSE, warning=FALSE, eval =FALSE}
# Removing some common ones
  #01100 - Metabolic Pathways
ko_path_list <- eggnogs %>%
  distinct(KEGG_ko, KEGG_Pathway) %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  filter(KEGG_ko != "-") %>%
  mutate(KEGG_Pathway = gsub("map|ko","", KEGG_Pathway)) %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  separate_rows(KEGG_Pathway, sep = ",") %>%
  distinct(KEGG_ko, KEGG_Pathway) %>%
  filter(!KEGG_Pathway %in% c("01100","01120")) 
```

```{r message=FALSE, warning=FALSE, eval=FALSE}
#removing pathways with only one obs
obs_in_path <- DE_genes %>%
  inner_join(ko_path_list, by = "KEGG_ko") %>%
  filter(type == "Down") %>%
  group_by(KEGG_Pathway) %>%
  tally() %>%
  ungroup() %>%
  filter(n >2)

#Copmbining the stupid fuucking list to try and do something 
DE_genes %>%
  inner_join(ko_path_list, by = "KEGG_ko") %>%
  filter(type == "Down") %>%
  inner_join(obs_in_path, by = "KEGG_Pathway")

vec <- DE_genes %>%
  filter(type == "Down") %>%
  distinct(KEGG_ko) %>% 
  pull(KEGG_ko)
  
```

```{r message=FALSE, warning=FALSE, eval=FALSE}
up <- read_tsv("../data/upreg_genes_TvsC.txt") %>%
  select(KO_term) %>%
  separate(KO_term, sep = 10, c("KEGG_ko","genename")) %>%
  mutate(KEGG_ko = gsub("ko:","", KEGG_ko)) %>%
  separate(genename, sep = ";", c("genename","desc"))
```
