---
title: "edgeR"
author: "Dennis Amnebrink"
date: "7/27/2022"
output:
  html_document:
      toc: yes
      toc_float:
        collapse: no
      fig_caption: yes
      code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(edgeR)
library(data.table)
library(viridis)
```

```{r file-loading, message=FALSE, warning=FALSE, cache=TRUE}
#File to connect NGI ID with our own classification ID, add the Âµ-values to this file for testing later, or do that in a sep file?
sample_ID <- read_tsv("../data/J.Pinhassi_21_02_sample_info.txt") %>%
  dplyr::select(1:2) %>%
  dplyr::rename(sample = 1, sample_name = 2) %>%
  mutate(sample_name = str_replace(sample_name, "[0-9]*:","")) %>%
  separate(sample_name, c("treatment","timepoint"), sep = ",") %>%
  mutate(treatment = gsub("NT","TN", treatment)) %>%
  mutate(sample_name = paste(timepoint, treatment, sep = "_")) %>%
  separate(treatment, c("treatment","replicate"), sep = -1) %>%
  mutate(tre_rep = gsub("_","", sample_name)) %>%
  mutate(tre_rep = toupper(tre_rep)) %>%
  mutate(day = case_when( timepoint == "t3" ~ 10,
                          TRUE ~ 17
                          )
         )

#Reading in annotations
eggnogs <- read_tsv("../data/eggnog_annotations.tsv.gz")

#Reading in taxonomic annotation of reads
taxonomy <- read_tsv("../data/eukulele_phylodb.tsv") %>%
  select(-species) %>%
  rename(species = "genus", genus = "family", family = "order", order = "class", class = "phylum", phylum = "kingdom")
  
#Count file for all orfs, tpm is calculated per sample
bbmap <- fread("../data/bbmap_counts.tsv.gz", sep = "\t") %>%
              filter(count > 0)  %>%                               # removing 0 counts to reduce size of table
              mutate(Geneid = str_replace(Geneid, "[0-9]*_",""),   # Joining in chr and geneid to match format of eggnog naming.
                     orf = paste(Chr,Geneid, sep = "_"),
                     sample = str_extract(sample, "P[0-9]*_[0-9]*")) %>% # Removing extra numbers in sample to match sample_ID format
              dplyr::select(-Geneid,-Chr) %>%
              dplyr::select(orf,Start,End,Strand,Length,sample,count,tpm) 
```

```{r constants, message=FALSE, warning=FALSE}
SIGNIFICANCE = 0.05
DIFF_TRESHOLD = 2.5
```

```{r R_colours}
GEOM_COL_COLOURS = c(
  RColorBrewer::brewer.pal(12, name = 'Paired'))
```

```{r prok-filter, message=FALSE, warning=FALSE, cache=TRUE}
#Removing non-prokreads as well as recalculating tpm over prok frac.
bbmap_p <- bbmap %>% 
  semi_join(taxonomy %>% filter(domain %in% c("Bacteria", "Archaea")), 
            by = "orf") %>%
  dplyr::select(-tpm) %>%
   group_by(sample) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
   ungroup()
```

# EdgeR-analysis day 10 
```{r edgeR_setup, message=FALSE, warning=FALSE}
# Do T vs C only day 10 (T3)
 bbmap_mat <- bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  inner_join(eggnogs %>% distinct(orf, KEGG_ko), by = "orf") %>%
  filter(KEGG_ko != "-") %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  group_by(sample, KEGG_ko) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "10") %>% #Removing day 10
  select(sample, KEGG_ko, count) %>%
  spread(sample, count, fill = 0) %>%
  column_to_rownames("KEGG_ko") %>%
  as.matrix()
 
library(edgeR)

y <- DGEList(bbmap_mat)

#Changing the grouping to the known treatments from mesocosm.
y$samples$group = sample_ID %>% filter(day == "10") %>% select(sample, treatment) %>%  pull(treatment)

#It has been correctly done as they match with the original sample file
groupF <- y$samples$group

#Another way of verifying
match(rownames(y$samples), colnames(bbmap_mat))

y <- DGEList(counts=bbmap_mat, group=groupF)

keep <- filterByExpr(y, group=groupF)

y <- y[keep, , keep.lib.sizes=FALSE]

y <- calcNormFactors(y)

y$samples

#The design-matrix doesn't like non-numeric categories, so here is a numeric corresponding to orignal levels, key =
# 1 = T, 2 = C, 3 = N, 4 = TN
design_m <- y$samples %>% select(group) %>%
  mutate(group = c(1,1,1,2,2,2,3,3,3,4,4,4))

y <- estimateDisp(y, design = design_m)

plotBCV(y)

#Creating a designmatrix (this looks like the correct structure), C is the base treatment, exlcuding intercept column.
trial_mat <- model.matrix(~0+groupF)

fit <- glmQLFit(y, trial_mat)

fit

#Comparing C vs T
colnames(trial_mat) = c("C","N","TN","t")

fit <- glmQLFit(y, trial_mat)

TvsC <- makeContrasts(t - C, levels=trial_mat)
NvsC <- makeContrasts(N - C, levels=trial_mat)
TNvsT <- makeContrasts(TN - t, levels=trial_mat)
```

```{r CvsT, message=FALSE, warning=FALSE}
qlf <- glmQLFTest(fit, contrast=TvsC)

#Most changing genes
topTags(qlf)

#Summary of genes up/downregulated, upregulated in this case refes to upregulated in C-treatment, negative is upregulated in TN-treatment.

summary(decideTests(qlf))
# Blue lines indicate 2-fold changes

pdf("../results/edgeRTNvsC2logFC.pdf")
plotMD(qlf)
abline(h=c(-1, 1), col="blue")
dev.off()

#Trying to look at the fold changes the edgeR way
tr <- glmTreat(fit, contrast=TvsC, lfc=log2(DIFF_TRESHOLD))

#The summary lets us know there are 1015 genes upregulated in  the C-treatment, and 4045 upregulated in  the TN treatment at a 2.5 logFC value.
diff_expr_sum <- summary(decideTestsDGE(tr)) %>% as.data.frame()

write_tsv(diff_expr_sum, "../results/edgeR_gene_partitioningCvsT.tsv")

TvsC <- tr$table %>%
  rownames_to_column("KEGG_ko")

#Doing the same but setting treshold to 2.5 logFC
pdf("../results/edgeRTvsC2logFC2.5.pdf")
plotMD(tr)
abline(h=c(-DIFF_TRESHOLD, DIFF_TRESHOLD), col="blue")
dev.off()
```

```{r NvsC, message=FALSE, warning=FALSE}
qlf <- glmQLFTest(fit, contrast=NvsC)

#Most changing genes
topTags(qlf)

#Summaroy of genes up/downregulated, upregulated in this case refes to upregulated in C-treatment, negative is upregulated in TN-treatment.

summary(decideTests(qlf))
# Blue lines indicate 2-fold changes

pdf("../results/edgeRCvsN2logFC.pdf")
plotMD(qlf)
abline(h=c(-1, 1), col="blue")
dev.off()

#Trying to look at the fold changes the edgeR way
tr <- glmTreat(fit, contrast=NvsC, lfc=log2(DIFF_TRESHOLD))

CvsN <- tr$table %>%
  rownames_to_column("KEGG_ko")
```

```{r TvsTN, message=FALSE, warning=FALSE}
qlf <- glmQLFTest(fit, contrast=TNvsT)

#Most changing genes
topTags(qlf)

#Summaroy of genes up/downregulated, upregulated in this case refes to upregulated in C-treatment, negative is upregulated in TN-treatment.

summary(decideTests(qlf))
# Blue lines indicate 2-fold changes

pdf("../results/edgeRTvsNT2logFC.pdf")
plotMD(qlf)
abline(h=c(-1, 1), col="blue")
dev.off()

#Trying to look at the fold changes the edgeR way
tr <- glmTreat(fit, contrast=TNvsT, lfc=log2(DIFF_TRESHOLD))

TvsTN <- tr$table %>%
  rownames_to_column("KEGG_ko")
```

```{r message=FALSE, warning=FALSE}
#Combiningputput df's and performing fdr correction
TvsC %>%
  as.data.frame() %>%
  mutate(contrast = "CvsT") %>%
  bind_rows(CvsN %>% mutate(contrast = "CsvN")) %>%
  bind_rows(TvsTN %>% mutate(contrast = "TvsTN")) %>%
  mutate(fdr = p.adjust(PValue, method = "fdr")) %>%
  mutate(type = 
           case_when(
             (fdr <= SIGNIFICANCE & logFC >= DIFF_TRESHOLD) ~ "Up",
             (fdr <= SIGNIFICANCE & logFC <= -DIFF_TRESHOLD) ~ "Down",
             TRUE ~"non-DE"
           )) %>%
  mutate(type = forcats::as_factor(type) %>% fct_relevel("Up","non-DE","Down")) %>%
  write_tsv("../results/edgeR_contrasts.tsv")
```

## Inspecting contrasts
```{r CvsT_KO, message=FALSE, warning=FALSE}
contrasts <- read_tsv("../results/edgeR_contrasts.tsv")

mat <- contrasts %>%
  filter(contrast == "CvsT") %>%
  filter(fdr <= SIGNIFICANCE) %>%
  arrange(desc(logFC)) %>%
  mutate(KEGG_ko = factor(KEGG_ko) %>% forcats::fct_reorder(logFC, .desc = TRUE)) %>%
  select(KEGG_ko, logFC) %>%
  spread(KEGG_ko, logFC, fill = 0) %>%
  as.matrix() 
 
 
 #use geom rug to add the colourstrip on the X axis
bbmap_plot <- bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  inner_join(eggnogs %>% distinct(orf, KEGG_ko), by = "orf") %>%
  filter(KEGG_ko != "-") %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  group_by(sample, KEGG_ko) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "10") %>%#Removing day 10 
  filter(treatment %in% c("C","T")) %>%
  group_by(treatment, KEGG_ko) %>%
  summarise(mean = mean(count), stdev = sd(count)) %>%
  ungroup() %>%
  inner_join(contrasts %>%
    filter(contrast == "CvsT") %>%
    filter(fdr <= SIGNIFICANCE) %>%
    arrange(desc(logFC)) %>%
    select(KEGG_ko, logFC), by = "KEGG_ko") %>%
  mutate(KEGG_ko = factor(KEGG_ko) %>% forcats::fct_reorder(logFC, .desc = TRUE)) 
  
bbmap_plot %>% 
  filter(mean <= 80000) %>%
  #filter(logFC>= 4 | logFC <= -4) %>%
 ggplot(mapping = aes(x = KEGG_ko, y = log(mean))) +
  geom_col( aes(fill = treatment)) +
  geom_rug(aes(colour = logFC), sides = "b") +
  scale_colour_viridis(discrete = FALSE) +
  theme(axis.text.x = element_blank()) +
  facet_wrap(~treatment)
  ggsave("../results/CvsT_KO_logFC.png")
```

## Verifying that genes are up in T when they are "down" - use this and put a heatmap of logFC underneath and order the genes based on logFC for a cool visualization
```{r TvsC_KO_plot, message=FALSE, warning=FALSE, fig.cap = "Differentially expressed genes C vs T"}

 DE_genes <- contrasts %>%
   filter(contrast == "TvsC") %>%
   filter(type %in% c("Up","Down"))

bbmap_p %>% 
  dplyr::select(-t, -tpm) %>%
  inner_join(eggnogs %>% distinct(orf, KEGG_ko), by = "orf") %>%
  filter(KEGG_ko != "-") %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  group_by(sample, KEGG_ko) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  inner_join(sample_ID, by = "sample") %>%
  filter(day == "10") %>% #Removing day 10
  select(sample, KEGG_ko, count, treatment) %>%
  filter(treatment %in% c("C","T")) %>%
  mutate(KEGG_ko = gsub("ko:","", KEGG_ko)) %>%
  semi_join(DE_genes %>% filter(type == "Down"), by = "KEGG_ko") %>%
  group_by(treatment, KEGG_ko) %>%
  summarise(count= mean(count)) %>%
  ungroup() %>%
  ggplot(mapping = aes(x = KEGG_ko, y = count, fill = treatment)) +
  geom_col()
```

## NvsC
```{r NvsC_KO, message=FALSE, warning=FALSE, eval=FALSE}
qlf <- glmQLFTest(fit, contrast=CvsN)

#Most changing genes
topTags(qlf)

#Summaroy of genes up/downregulated, upregulated in this case refes to upregulated in C-treatment, negative is upregulated in TN-treatment.

summary(decideTests(qlf))
# Blue lines indicate 2-fold changes

pdf("../results/edgeRNvsC2logFC.pdf")
plotMD(qlf)
abline(h=c(-1, 1), col="blue")
dev.off()

#Trying to look at the fold changes the edgeR way
tr <- glmTreat(fit, contrast=NvsC, lfc=log2(DIFF_TRESHOLD))

#The summary lets us know there are 1015 genes upregulated in  the C-treatment, and 4045 upregulated in  the TN treatment at a 2.5 logFC value.
diff_expr_sum <- summary(decideTestsDGE(tr)) %>% as.data.frame()

write_tsv(diff_expr_sum, "../results/edgeR_gene_partitioningNvsC.tsv")

#Does the fdr correction apply to the summary?
obj <- tr$table %>%
  #mutate(fdr = p.adjust(PValue, method = "fdr")) %>%
  mutate(type = 
           case_when(
             (fdr <= SIGNIFICANCE & logFC >= DIFF_TRESHOLD) ~ "Up",
             (fdr <= SIGNIFICANCE & logFC <= -DIFF_TRESHOLD) ~ "Down",
             TRUE ~"non-DE"
           )) %>%
  mutate(type = forcats::as_factor(type) %>% fct_relevel("Up","non-DE","Down")) %>%
  group_by(type) %>%
  tally() %>%
  ungroup() # a few genes are missing using this approach, in the significant fractions, but are not in the non-DE?? The total genes still add up using the summary(decideTests(tr)), on the original edgeR-object...

CvsN <- tr$table %>%
  rownames_to_column("KEGG_ko")

#Doing the same but setting treshold to 2.5 logFC
pdf("../results/edgeRNvsC2logFC2.5.pdf")
plotMD(tr)
abline(h=c(-DIFF_TRESHOLD, DIFF_TRESHOLD), col="blue")
dev.off()
```

```{r fdr_correction, message=FALSE, warning=FALSE, eval=FALSE}

```

## Creating a key with pathway and KO-term
```{r message=FALSE, warning=FALSE, eval =FALSE}
# Removing some common ones
  #01100 - Metabolic Pathways
ko_path_list <- eggnogs %>%
  distinct(KEGG_ko, KEGG_Pathway) %>%
  separate_rows(KEGG_ko, sep = ",") %>%
  filter(KEGG_ko != "-") %>%
  mutate(KEGG_Pathway = gsub("map|ko","", KEGG_Pathway)) %>%
  mutate(KEGG_ko = gsub("ko:","",KEGG_ko)) %>%
  separate_rows(KEGG_Pathway, sep = ",") %>%
  distinct(KEGG_ko, KEGG_Pathway) %>%
  filter(!KEGG_Pathway %in% c("01100","01120")) 
```

```{r message=FALSE, warning=FALSE, eval=FALSE}
#removing pathways with only one obs
obs_in_path <- DE_genes %>%
  inner_join(ko_path_list, by = "KEGG_ko") %>%
  filter(type == "Down") %>%
  group_by(KEGG_Pathway) %>%
  tally() %>%
  ungroup() %>%
  filter(n >2)

#Copmbining the stupid fuucking list to try and do something 
DE_genes %>%
  inner_join(ko_path_list, by = "KEGG_ko") %>%
  filter(type == "Down") %>%
  inner_join(obs_in_path, by = "KEGG_Pathway")

vec <- DE_genes %>%
  filter(type == "Down") %>%
  distinct(KEGG_ko) %>% 
  pull(KEGG_ko)
  
```

```{r message=FALSE, warning=FALSE, eval=FALSE}
up <- read_tsv("../data/upreg_genes_TvsC.txt") %>%
  select(KO_term) %>%
  separate(KO_term, sep = 10, c("KEGG_ko","genename")) %>%
  mutate(KEGG_ko = gsub("ko:","", KEGG_ko)) %>%
  separate(genename, sep = ";", c("genename","desc"))
```
