---
title: "PROMAC_MetaT"
author: "Dennis Amnebrink"
date: "`r format(Sys.time(), '%Y-%m-%d')`"
output:
  html_document:
      toc: yes
      toc_float:
        collapse: no
      fig_caption: yes
      code_folding: hide
---
# A higher abundance of proteorhodopsin-genes in the N-treatment? Based on 16S data....
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries}
#Loading libraries
library(tidyverse)
library(vegan)
library(pheatmap)
library(data.table)
```

```{r R_colours}
GEOM_COL_COLOURS = c(
  RColorBrewer::brewer.pal(12, name = 'Paired'))
```

```{r constants}
SIGNIFICANCE = 0.05
DIFF_TRESHOLD = 2.5
```

```{r files}
#File to connect NGI ID with our own classification ID, add the µ-values to this file for testing later, or do that in a sep file?
sample_ID <- read_tsv("../data/J.Pinhassi_21_02_sample_info.txt") %>%
  dplyr::select(1:2) %>%
  dplyr::rename(sample = 1, sample_name = 2) %>%
  mutate(sample_name = str_replace(sample_name, "[0-9]*:","")) %>%
  separate(sample_name, c("treatment","timepoint"), sep = ",") %>%
  mutate(treatment = gsub("NT","TN", treatment)) %>%
  mutate(sample_name = paste(timepoint, treatment, sep = "_")) %>%
  separate(treatment, c("treatment","replicate"), sep = -1) %>%
  mutate(tre_rep = gsub("_","", sample_name)) %>%
  mutate(tre_rep = toupper(tre_rep)) %>%
  mutate(day = case_when( timepoint == "t3" ~ 10,
                          TRUE ~ 17
                          )
         )

#Reading in annotations
eggnogs <- read_tsv("../data/eggnog_annotations.tsv.gz")

#Reading in taxonomic annotation of reads
taxonomy <- read_tsv("../data/eukulele_phylodb.tsv") %>%
  select(-species) %>%
  rename(species = "genus", genus = "family", family = "order", order = "class", class = "phylum", phylum = "kingdom")
  
#Count file for all orfs, tpm is calculated per sample
bbmap <- fread("../data/bbmap_counts.tsv.gz", sep = "\t") %>%
              filter(count > 0)  %>%                               # removing 0 counts to reduce size of table
              mutate(Geneid = str_replace(Geneid, "[0-9]*_",""),   # Joining in chr and geneid to match format of eggnog naming.
                     orf = paste(Chr,Geneid, sep = "_"),
                     sample = str_extract(sample, "P[0-9]*_[0-9]*")) %>% # Removing extra numbers in sample to match sample_ID format
              dplyr::select(-Geneid,-Chr) %>%
              dplyr::select(orf,Start,End,Strand,Length,sample,count,tpm) 

#Overarching stats for all sample files
overall_stats <- read_tsv("../data/overall_stats.tsv") %>%
  dplyr::select(-2,-3)                                                   # Removing empty columns

target_genes <- read_tsv("../data/list_of_genes.tsv")

mu_values <- read_tsv("../data/mu.tsv")

#Add maintenance respiration values
Rm <- read_tsv("../data/Rm_values.tsv") %>%
   mutate(tre_rep = paste(timepoint, treatment, replicate, sep = "")) 


```

```{r read_overview}
#Add one with taxonomic annotation too, and one with original reads
  overall_stats %>% 
    gather(key = "process", value = "value", c(2:4)) %>%
    mutate(process = as_factor(process)) %>%
    ggplot(mapping = aes(x = process, y = value, group = process, fill = process)) +
    geom_boxplot() +
    ggtitle("Boxplot distribution of reads throughout pipeline")
```

## Plotting the number of reads to each domain to characterize community
```{r ASV-domain-fidelity}
taxonomy %>%
  group_by(domain) %>%
  tally() %>%
  ungroup() %>%
  ggplot(mapping = aes(x = domain, y = n, group = domain, fill = domain)) +
  geom_col()
```

```{r ASV-domain-abundance}
bbmap %>%
  inner_join(taxonomy, by = "orf") %>%
  group_by(domain) %>%
  summarise(count = sum(count)) %>%
  ungroup() %>%
  mutate(percent = count/sum(count)) %>%
  ggplot(mapping = aes(x = domain, y = percent, group = domain, fill = domain)) +
  geom_col() +
  theme(axis.text.x = element_text(face = c("plain","bold","plain","plain","plain")))
```

# Removing non-prokaryotic reads
```{r}
#Removing non-prokreads as well as recalculating tpm over prok frac.
bbmap_p <- bbmap %>% 
  semi_join(taxonomy %>% filter(domain %in% c("Bacteria", "Archaea")), 
            by = "orf") %>%
  dplyr::select(-tpm) %>%
   group_by(sample) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
   ungroup()
```

## COG-nMDS
```{r cog-cat-overview, eval=FALSE}
#COG-categories shoed no pattern, they are very large as well, the next chunk shows clear patterns (higher resolution)
orf_mat <- bbmap_p %>%
  inner_join(eggnogs, by = "orf") %>% #6674350 rows
  inner_join(sample_ID, by = "sample") %>%
  select(best_og_cat, tpm, sample_name) %>%
  separate_rows(best_og_cat, sep = '(?<=.)(?=.)') %>% # Not sure how this separator works, but it works
  group_by(sample_name, best_og_cat) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  spread(best_og_cat, tpm, fill = 0) %>%
  column_to_rownames("sample_name") %>%
  as.matrix() 

  orf_mat_hellinger <- decostand(orf_mat[,-1], method = "hellinger") # Removing the unannotated functional category and doing helllinger
  
  nmds <- metaMDS(orf_mat[,-1], distance = "bray", k = 2) 
  nmds <- metaMDS(orf_mat_hellinger, distance = "bray", k = 2) 
  
  
  as.data.frame(nmds$points) %>%
    rownames_to_column("sample_name") %>%
    inner_join(sample_ID, by = "sample_name") %>%
    #separate(treatment, c("treatment", "replicate"), sep = -1) %>%
    ggplot(mapping = aes(x = MDS1, y = MDS2, fill = treatment, colour = treatment, shape = timepoint)) +
    geom_point() +
    scale_colour_manual(values = GEOM_COL_COLOURS) +
    ggtitle("nMDS based on COG-categories") +
    ylab("NMDS2") +
    xlab(paste0("NMDS1\n", "stress=", round(nmds$stress, 3))) 
    ggsave("../results/NMDS_COG_CATS.png")
    
  PCA <- rda(orf_mat[,-1], distance = "bray") 
  PCA <- rda(orf_mat_hellinger, distance = "bray") 
  
   eigenvalues <- as.data.frame(PCA$CA$eig) %>%
    rownames_to_column("PC_axes") %>%
    rename(eigenvalues = "PCA$CA$eig") %>%
    mutate(eigenvalues_variation_explained = (eigenvalues/sum(eigenvalues) * 100))
   
   as.data.frame(PCA$CA$u) %>%
     rownames_to_column("sample_name") %>%
     inner_join(sample_ID, by = "sample_name") %>%
     #separate(treatment, c("treatment", "replicate"), sep = -1) %>%
     ggplot(mapping = aes(x = PC1, y = PC2, fill = treatment, colour = treatment, shape = timepoint)) +
     geom_point(aes(size = 10)) +
     scale_colour_manual(values = GEOM_COL_COLOURS) +
     xlab(paste("PC1", round(eigenvalues$eigenvalues_variation_explained[1], digits = 2),"%", sep = " ")) +
     ylab(paste("PC2", round(eigenvalues$eigenvalues_variation_explained[2], digits = 2),"%", sep = " ")) 
     ggtitle("PCA based on COG-categories") 
     ggsave("../results/PCA_COG_CATS.png")
```

## eggNOG-PCA
```{r eggNOG-overview cache=TRUE}
   # Doing nmds and PCA on eggnogs instead, Looks a lot better
orf_mat <-  bbmap_p %>%
  inner_join(eggnogs, by = "orf") %>% #6674350 rows
  inner_join(sample_ID, by = "sample") %>%
  rename(eggNOG_OGs = "eggNOG OGs") %>%
  select(eggNOG_OGs, tpm, sample_name) %>%
  separate_rows(eggNOG_OGs, sep = ',') %>% 
  group_by(sample_name, eggNOG_OGs) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  spread(eggNOG_OGs, tpm, fill = 0) %>%
  column_to_rownames("sample_name") %>%
  as.matrix() 
   
   orf_mat_hellinger <- decostand(orf_mat[,-1], method = "hellinger") # Removing the unannotated functional category and doing helllinger¨
     
    nmds <- metaMDS(orf_mat[,-1], distance = "bray", k = 2) 
   
  
  as.data.frame(nmds$points) %>%
    rownames_to_column("sample_name") %>%
    inner_join(sample_ID, by = "sample_name") %>%
    #separate(treatment, c("treatment", "replicate"), sep = -1) %>%
    ggplot(mapping = aes(x = MDS1, y = MDS2, fill = treatment, colour = treatment, shape = timepoint)) +
    geom_point() +
    scale_colour_manual(values = GEOM_COL_COLOURS) +
    ggtitle("nMDS based on eggNOGS") +
    ylab("NMDS2") +
    xlab(paste0("NMDS1\n", "stress=", round(nmds$stress, 3))) 
    ggsave("../results/NMDS_eggNOG.png")
    
    
  nmds <- metaMDS(orf_mat_hellinger, distance = "bray", k = 2) 
     
  as.data.frame(nmds$points) %>%
    rownames_to_column("sample_name") %>%
    inner_join(sample_ID, by = "sample_name") %>%
    #separate(treatment, c("treatment", "replicate"), sep = -1) %>%
    ggplot(mapping = aes(x = MDS1, y = MDS2, fill = treatment, colour = treatment, shape = timepoint)) +
    geom_point() +
    scale_colour_manual(values = GEOM_COL_COLOURS) +
    ggtitle("nMDS based on eggNOGS") +
    ylab("NMDS2") +
    xlab(paste0("NMDS1\n", "stress=", round(nmds$stress, 3))) 
    ggsave("../results/NMDS_hellinger_eggNOG.png")
  
  
  PCA <- rda(orf_mat_hellinger, distance = "bray") 
  
   eigenvalues <- as.data.frame(PCA$CA$eig) %>%
    rownames_to_column("PC_axes") %>%
    rename(eigenvalues = "PCA$CA$eig") %>%
    mutate(eigenvalues_variation_explained = (eigenvalues/sum(eigenvalues) * 100))
   
   as.data.frame(PCA$CA$u) %>%
     rownames_to_column("sample_name") %>%
     inner_join(sample_ID, by = "sample_name") %>%
     #separate(treatment, c("treatment", "replicate"), sep = -1) %>%
     ggplot(mapping = aes(x = PC1, y = PC2, fill = treatment, colour = treatment, shape = timepoint)) +
     geom_point() +
     scale_colour_manual(values = GEOM_COL_COLOURS) +
     xlab(paste("PC1", round(eigenvalues$eigenvalues_variation_explained[1], digits = 2),"%", sep = " ")) +
     ylab(paste("PC2", round(eigenvalues$eigenvalues_variation_explained[2], digits = 2),"%", sep = " ")) +
     ggtitle("PCA based on eggNOGS") +
     theme_minimal() +
     theme(panel.grid = element_blank()) +
      geom_vline(xintercept = 0, linetype = "dashed", size = 0.5) +
     geom_hline(yintercept=0, linetype="dashed", size = 0.5) 
     ggsave("../results/PCA_hellinger_eggNOGS.png")
```

# Stacked bars for visualization of taxonomy
```{r}
#Create taxonomy layout with top10 phyla, or class or order for example.
top12 <- bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  group_by(treatment, timepoint, orf) %>%
  summarise(tpm = mean(tpm)) %>%
  ungroup() %>%
  inner_join(taxonomy, by = "orf") %>%
  group_by(phylum) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  arrange(desc(tpm)) %>%
  slice(1:12) %>% view()
  pull(phylum)

# Line for mutating in the new annotation using 12 families only, and rest in other category
#mutate(Group = ifelse(grepl(paste(top10, collapse = "|"), family), paste0(family), "other"))

 #Adding one more colour to the scale
  GEOM_COL_COLOURS1 = c(GEOM_COL_COLOURS, "#000000")
  
 bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  dplyr::select(-tpm) %>%
  group_by(treatment,timepoint) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
            ungroup() %>%
  inner_join(taxonomy, by = "orf") %>%
  mutate(Group = ifelse(phylum %in% top12, paste0(phylum), "other")) %>%
  group_by(treatment, timepoint, Group) %>% #Summarising mean tpm per timepoint and treatment per group
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  ggplot(mapping = aes(x= timepoint, y = tpm, fill = fct_relevel(Group, "other", after = 10))) +
  geom_col() +
  facet_wrap(~ treatment) +
  theme(axis.text.x = element_text(angle = 60), panel.background = element_blank()) +
  scale_fill_manual('phylum', values = GEOM_COL_COLOURS1) +
  ylab("Relative abundance") +
  xlab("Day") +
  ggtitle("taxnomy of reads, phylum level") 
  ggsave("../results/taxonomy_phylum.png")
  

#Plotting again but with class
 top12 <- bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  group_by(treatment, timepoint, orf) %>%
  summarise(tpm = mean(tpm)) %>%
  ungroup() %>%
  inner_join(taxonomy, by = "orf") %>%
  group_by(class) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  arrange(desc(tpm)) %>%
  slice(1:12) %>%
  pull(class)
 
bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  dplyr::select(-tpm) %>%
  group_by(treatment,timepoint) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
            ungroup() %>%
  inner_join(taxonomy, by = "orf") %>%
  mutate(Group = ifelse(class %in% top12, paste0(class), "other")) %>%
  group_by(treatment, timepoint, Group) %>% #Summarising mean tpm per timepoint and treatment per group
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  ggplot(mapping = aes(x= timepoint, y = tpm, fill = fct_relevel(Group, "other", after = 10))) +
  geom_col() +
  facet_wrap(~ treatment) +
  theme(axis.text.x = element_text(angle = 60), panel.background = element_blank()) +
  scale_fill_manual('class', values = GEOM_COL_COLOURS1) +
  ylab("Relative abundance") +
  xlab("Day") +
  ggtitle("Taxonomy of reads, class level") 
  ggsave("../results/taxonomy_class.png")

#Plotting again but with order
 top12 <- bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  group_by(treatment, timepoint, orf) %>%
  summarise(tpm = mean(tpm)) %>%
  ungroup() %>%
  inner_join(taxonomy, by = "orf") %>%
  group_by(order) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  arrange(desc(tpm)) %>%
  slice(1:12) %>%
  pull(order)
 
bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  dplyr::select(-tpm) %>%
  group_by(treatment,timepoint) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
            ungroup() %>%
  inner_join(taxonomy, by = "orf") %>%
   mutate(Group = ifelse(order %in% top12, paste0(order), "other")) %>%
  group_by(treatment, timepoint, Group) %>% #Summarising mean tpm per timepoint and treatment per group
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  ggplot(mapping = aes(x= timepoint, y = tpm, fill = fct_relevel(Group, "other", after = 10))) +
  geom_col() +
  facet_wrap(~ treatment) +
  theme(axis.text.x = element_text(angle = 60), panel.background = element_blank()) +
  scale_fill_manual('order', values = GEOM_COL_COLOURS1) +
  ylab("Relative abundance") +
  xlab("Day") +
  ggtitle("Taxonomy of reads, order level") 
  ggsave("../results/taxonomy_order.png")

#Plotting again but with family
 top12 <- bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  group_by(treatment, timepoint, orf) %>%
  summarise(tpm = mean(tpm)) %>%
  ungroup() %>%
  inner_join(taxonomy, by = "orf") %>%
  group_by(family) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  arrange(desc(tpm)) %>%
  slice(1:12) %>%
  .$family
 
bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  dplyr::select(-tpm) %>%
  group_by(treatment,timepoint) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
            ungroup() %>%
  inner_join(taxonomy, by = "orf") %>%
   mutate(Group = ifelse(family %in% top12, paste0(family), "other")) %>%
  group_by(treatment, timepoint, Group) %>% #Summarising mean tpm per timepoint and treatment per group
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  ggplot(mapping = aes(x= timepoint, y = tpm, fill = fct_relevel(Group, "other", after = 10))) +
  geom_col() +
  facet_wrap(~ treatment) +
  theme(axis.text.x = element_text(angle = 60), panel.background = element_blank()) +
  scale_fill_manual('family', values = GEOM_COL_COLOURS1) +
  ylab("Relative abundance") +
  xlab("Day") +
  ggtitle("taxonomy of reads, family level") 
  ggsave("../results/taxonomy_family.png")
```

## Family level taxonomy expression within phyla

### Bacteroidetes/Chlorobi group

Should I perhaps lump together the unclassified Flavobacteriales with the NA?
```{r}
#Family level expression in Bacteroidetes/Chlorobi group. total of 23 families, look at top12

#I need to set the variables, 1. taxonomic level (e.g phylum, order), 2. Group to summarise on e.g bacteroidetes/chlorobi group. 

tax_magnify <- function(taxlevel = "phylum", taxdf = taxonomy,  g = "Bacteroidetes/Chlorobi group", sampledf = sample_ID) {
  
  top11 <- bbmap_p %>%
    inner_join(taxdf, by= "orf") %>%
    filter(.[[taxlevel]] == g) %>%
    inner_join(sampledf, by = "sample") %>%
    group_by(treatment,timepoint) %>%
              mutate(t = count/Length) %>%
              mutate(tpm = t/sum(t)*1e6) %>%
              ungroup() %>%
    group_by(family) %>%
    summarise(tpm = sum(tpm)) %>%
    ungroup() %>%
    distinct(family, tpm) %>%
    arrange(desc(tpm)) %>%
    slice(1:11) %>%
    pull(family)
  
  
  bbmap_p %>%
    inner_join(taxdf, by= "orf") %>%
    filter(.[[taxlevel]] == g) %>%
    inner_join(sampledf, by = "sample") %>%
    group_by(treatment,timepoint) %>%
              mutate(t = count/Length) %>%
              mutate(tpm = t/sum(t)*1e6) %>%
              ungroup() %>%
    mutate(n_family = ifelse(family %in% top11, paste(family), "Other")) %>%
    group_by(n_family, treatment, timepoint) %>%
    summarise(tpm = sum(tpm)) %>%
    ungroup() %>%
    ggplot(mapping = aes(x = timepoint, y = tpm, fill = n_family)) +
    geom_col() +
    facet_wrap(~ treatment) +
    scale_fill_manual(values = GEOM_COL_COLOURS, na.value = "#000000") +
    ggtitle(g)

}
```

```{r bacteroidetes_chlorobi_families}
tax_magnify()
```

### Alphaproteobacteria
The proteobacterial expression was mainly explained by Alpha, Beta- and Gammaproteobacteria, with the highest fraction of Alphaproteobacteria in the T-treatment. Gammaproteobacteria was more common in the TN-treatment. Roughly half of the  alphaproteobacterial expression was derived from Rhodobacteraceae and Surface 1-category, the latter being part of the SAR11-clade and constituted by unclassified bacteria or the genus Pelagibacter sp.. The expression composition was similar across all treatments, with the tendency of higher expression by unknown Alphaproteobacteria at day 17 in the TN treatment.
```{r}
tax_magnify(taxlevel = "class", g = "Alphaproteobacteria")
```

### Gammaproteobacteria
In Gammaproteobacteria, expression composition was markedly different in the N-treatment as more than 25% of expression was explained by *Shewanellaceae* at both timepoints. In the TN-treatment Oceanospirillaceae was instead the most prominent during day 10 while expression was more even during day 17. The C-treatment showed the most diverse expression composition with the larger groups comprising Environmental Gammaproteobacteria and NA. 
```{r}
tax_magnify(taxlevel = "class", g = "Gammaproteobacteria")
```

### Thaumarchaeota
The expression was represented mainly by undetermined thaumarchaeota families, of which the Genus *Nitrosopumilus sp.* was a member, a small fraction was attributed to the Cenarchaeaceae family.
```{r}
tax_magnify(taxlevel = "phylum", g = "Thaumarchaeota")
```

# Kegg-pathways stacked bars
This gave nothing of real insight, or interesting patterns. Non-hierarchical structure doesn't help.
```{r eval=FALSE}
#Getting the most abundant categories
top12Path <- bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  dplyr::select(-tpm) %>%
  group_by(treatment,timepoint) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
            ungroup() %>%
  select(-t) %>%
  inner_join(eggnogs, by = "orf") %>%
  mutate(KEGG_Pathway = gsub("[a-z]*","", KEGG_Pathway)) %>% #Remove ko and map-prefix to avoid duplication.
  separate_rows(KEGG_Pathway, sep = ",") %>%
  distinct(KEGG_Pathway, treatment, timepoint, orf, tpm) %>% #Getting rid of duplicates due to the removal of prefixes
  group_by(KEGG_Pathway) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  arrange(desc(tpm)) %>%
  filter(KEGG_Pathway!= "-") %>%
  slice(1:12) %>%
  pull(KEGG_Pathway)

bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  dplyr::select(-tpm) %>%
  group_by(treatment,timepoint) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
            ungroup() %>%
  select(-t) %>%
  inner_join(eggnogs, by = "orf") %>%
  mutate(KEGG_Pathway = gsub("[a-z]*","", KEGG_Pathway)) %>% #Remove ko and map-prefix to avoid duplication.
  separate_rows(KEGG_Pathway, sep = ",") %>%
  distinct(KEGG_Pathway, treatment, timepoint, orf, tpm) %>% #Getting rid of duplicates due to the removal of prefixes
  mutate(Group = ifelse(KEGG_Pathway %in% top12Path, paste0(KEGG_Pathway), "Other")) %>%
  group_by(treatment, timepoint, Group) %>% #Summarising mean tpm per timepoint and treatment per group
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  ggplot(mapping = aes(x= timepoint, y = tpm, fill = fct_relevel(Group, "Other", after = 12))) +
  geom_col() +
  facet_wrap(~ treatment) +
  theme(axis.text.x = element_text(angle = 60), panel.background = element_blank()) +
  scale_fill_manual('KEGG Pathway', values = GEOM_COL_COLOURS1) +
  ylab("Relative abundance") +
  xlab("Day")
  ggsave("../results/KEGG_paths.png")
  ggsave("../results/KEGG_paths.pdf")
```

# TOP50 genes
```{r top_genes}
annotated_counts <- bbmap_p %>%
  inner_join(eggnogs, by = "orf") %>% #6674350 rows
  inner_join(sample_ID, by = "sample") 

#Top50 genes overall are not stress genes, very good sign! Rather ammonium transport, photosystem, flagellar motorproteins, and others.
  

top50all <- annotated_counts %>%
  group_by(orf) %>%
  summarise(tpm = mean(tpm)) %>%
  ungroup() %>%
  arrange(desc(tpm)) %>%
  slice(1:50) %>% 
  inner_join(eggnogs, by = "orf") %>%
  arrange(desc(tpm)) %>%
  pull(orf)
  

#Top genes per treatment
top50 <- annotated_counts %>%
  group_by(treatment, orf) %>%
  summarise(tpm = mean(tpm)) %>%
  ungroup() %>%
  group_by(treatment) %>%
  arrange(desc(tpm)) %>%
  top_n(50, tpm) %>% 
  ungroup() %>%
  inner_join(eggnogs, by = "orf") %>% 
  select(orf, Preferred_name, best_og_desc, treatment) 

#Plotting the genes, first in a venn diagram
#Creating vectores with the top15 asv's per treatment
C_orf <- top50 %>%
  filter(treatment == "C") %>%
  .$orf

T_orf <- top50 %>%
  filter(treatment == "T") %>%
  .$orf

N_orf <- top50 %>%
  filter(treatment == "N") %>%
  .$orf

TN_orf <- top50 %>%
  filter(treatment == "NT") %>%
  .$orf

venn.diagram(
  x = list(C_orf, N_orf, T_orf, TN_orf),
  category.names = c("C" , "N", "T","TN"),
  filename = '../results/top50_shared.png',
  output=TRUE
)

# Interestingly not as many shared between all treatments, and quite a few unique genes for the T-treatments
```

## Plotting the abundance of the 50 most abundant genes separated per treatment to see contribution
```{r}
top50 <-  bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  dplyr::select(-tpm) %>%
  group_by(treatment,timepoint) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
            ungroup() %>%
  group_by(orf) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  arrange(desc(tpm)) %>%
  slice(1:50) %>%
  pull(orf)

#Plotting the top50 genes
plot_ibj <- bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  dplyr::select(-tpm) %>%
  group_by(treatment,timepoint) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
            ungroup() %>%
  group_by(orf, treatment, timepoint) %>%
  summarise(tpm = mean(tpm)) %>%
  ungroup() %>%
  left_join(taxonomy, by ="orf") 

  p <- plot_ibj %>%
    filter(orf %in% top50) %>%
    ggplot(mapping = aes(x = orf, y = tpm, fill = treatment)) +
    geom_col(position = "dodge") +
    theme(axis.text.x = element_text(hjust = 1, angle = 60)) +
    facet_wrap(~ timepoint)
  
  ggplotly(p)
  
  #Trying the pheatmap way to investigate the top50 genes, and their annotation
  top50_matr <- plot_ibj %>%
    filter(orf %in% top50) %>%
    mutate(treatpoint = paste(treatment, timepoint, sep = "_")) %>%
    select(orf,treatpoint, tpm) %>%
    spread(treatpoint, tpm, fill = 0) %>%
    column_to_rownames("orf") %>%
    as.matrix()
  
  #MAKE TREATMENT VECTOR
Treatment <- c("C","C","N","N","T","T","TN","TN")


#Connecting the annotation column by adding rownames to match colnames in matrix
ann_col <- data.frame(Treatment)
rownames(ann_col) <- colnames(top50_matr)
col_vec <- rep(c("10","15"), times = 4)

my_colour = list(Treatment =c("C" = "blue", "N" = "green3", "T" = "red", "TN" = "yellow4")
)

#Lastly add annotation fore the rows based on eggnog-annotation.
row_names <- top50_matr %>%
  as.data.frame() %>%
  rownames_to_column("orf") %>%
  left_join(eggnogs, by = "orf") %>%
  select(orf, Preferred_name) %>%
  column_to_rownames("orf") %>%
  pull(Preferred_name)

top50_matr %>%
  as.data.frame() %>%
  rownames_to_column("orf") %>%
  left_join(eggnogs, by = "orf") %>%
  select(orf, Preferred_name) %>%
  column_to_rownames("orf") %>% view()

#Add zero values in a different colour for identifying absence of genes
  
pheatmap(top50_matr, cluster_cols = F, clustering_distance_rows = "correlation",annotation_col = ann_col, labels_col = col_vec, labels_row = row_names ,gaps_col = c(2,4,6,8), annotation_colors = my_colour)

p2 <- pheatmap(top50_matr, cluster_cols = F, clustering_distance_rows = "correlation", annotation_col = ann_col, scale = "row",labels_col = col_vec, labels_row = row_names ,gaps_col = c(2,4,6,8), annotation_colors = my_colour, treeheight_row = 0, fontsize = 5, border_color = NA)

p1 <- pheatmap(top50_matr, cluster_cols = F, clustering_distance_rows = "correlation",annotation_col = ann_col, labels_col = col_vec, labels_row = row_names, show_rownames = F ,gaps_col = c(2,4,6,8), annotation_colors = my_colour,annotation_names_col = F, annotation_legend = F, fontsize =5, border_color = NA)

# Out of the 50 most abundant genes, the majority are unidentified, the motility response is very strong in the TN treatment as seen by the expression of the fliM gene, which is also the most abundant gene in the C treatment. A suite of unknown genes are also representative for the C-treatment as they for a very distinct cluster.


pdf("../results/top50_both.pdf")
grid.arrange(grobs = list(p1[[4]],
                          p2[[4]]
), ncol = 2)

grid.arrange(grobs = list(p1[[4]],
                          p2[[4]]
), ncol = 2)
dev.off()
```

## Most abundant KEGG-modules across all treatments
```{r}
top50 <-  bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  dplyr::select(-tpm) %>%
  group_by(treatment,timepoint) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
            ungroup() %>%
  inner_join(eggnogs, by = "orf") %>%
  separate_rows(KEGG_Module, sep = ",") %>%
  group_by(KEGG_Module) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  filter(KEGG_Module != "-") %>%
  arrange(desc(tpm)) %>%
  slice(1:50) %>%
  pull(KEGG_Module)

top50

#Module legend
  #M00178 - No entry
  #M00179 - No entry
  #M00157 - F-type ATPase, prokaryotes and chloroplast
  #M00237 - No entry
  #M00804 - Complete nitrification, comammox, ammonia => nitrite => nitrate
  #M00174 - Methane oxidation, methanotroph, methane => formaldehyde
  #M00528 - Nitrification, ammonia => nitrite
  #M00222 - No entry
  #M00183 - No entry
  #M00009 - Citrate cycle (TCA cycle, Krebs cycle)
  #M00335 - No entry
  #M00011 - Citrate cycle, second carbon oxidation, 2-oxoglutarate => oxaloacetate
  #M00173 - Reductive citrate cycle (Arnon-Buchanan cycle)
  #M00083 - Fatty acid biosynthesis, elongation
  #M00529 - Denitrification, nitrate => nitrogen
  #M00374 - Dicarboxylate-hydroxybutyrate cycle
  #M00360 - No entry
  #M00359 - No entry
  #M00232 - No entry 
  #M00082 - Fatty acid biosynthesis, initiation
  #M00012 - Glyoxylate cycle
  #M00570 - Isoleucine biosynthesis, threonine => 2-oxobutanoate => isoleucine
  #M00019 - Valine/isoleucine biosynthesis, pyruvate => valine / 2-oxobutanoate => isoleucine
  #M00053 - Deoxyribonucleotide biosynthesis, ADP/GDP/CDP/UDP => dATP/dGTP/dCTP/dUTP
  #M00048 - De novo purine biosynthesis, PRPP + glutamine => IMP
  #M00742 - No entry
  #M00144 - NADH:quinone oxidoreductase, prokaryotes
  #M00155 - Cytochrome c oxidase, prokaryotes
  #M00087 - beta-Oxidation
  #M00003 - Gluconeogenesis, oxaloacetate => fructose-6P
  #M00376 - 3-Hydroxypropionate bi-cycle
  #M00532 - Photorespiration
  #M00373 - Ethylmalonyl pathway
  #M00001 - Glycolysis (Embden-Meyerhof pathway), glucose => pyruvate
  #M00017 - Methionine biosynthesis, aspartate => homoserine => methionine
  #M00572 - Pimeloyl-ACP biosynthesis, BioC-BioH pathway, malonyl-ACP => pimeloyl-ACP
  #M00032 - Lysine degradation, lysine => saccharopine => acetoacetyl-CoA
  #M00740 - Methylaspartate cycle
  #M00620 - Incomplete reductive citrate cycle, acetyl-CoA => oxoglutarate
  #M00036 - Leucine degradation, leucine => acetoacetate + acetyl-CoA
  #M00172 - C4-dicarboxylic acid cycle, NADP - malic enzyme type
  #M00169 - CAM (Crassulacean acid metabolism), light
  #M00016 - Lysine biosynthesis, succinyl-DAP pathway, aspartate => lysine
  #M00346 - Formaldehyde assimilation, serine pathway
  #M00236 - No entry
  #M00165 - Reductive pentose phosphate cycle (Calvin cycle)
  #M00307 - Pyruvate oxidation, pyruvate => acetyl-CoA
  #M00151 - Cytochrome bc1 complex respiratory unit
  #M00152 - Cytochrome bc1 complex
  #M00394 - No entry.

# While a lot of the abundant modules have no entry in the database. Among the most abundant modules are the over-represented categories in nitrogen metabolism, along with denitrification and methane oxidation (most likely ammonia oxidation, but not necessarily). Other abundant categories included the TCA-cycle, reductive TCA-cycle,an ATPase and fatty acid biosynthesis (elongation). Interestingly, since not many were found to be over-represented, instead characteristics for the entire community included modules in Nitrogen metabolism, such as comammox, denitrification, and  dissimilatory nitrate reduction.

#Create table with top 50, and annotate whether or not they are over-represented
bbmap_p %>%
  inner_join(sample_ID, by = "sample") %>%
  dplyr::select(-tpm) %>%
  group_by(treatment,timepoint) %>%
            mutate(t = count/Length) %>%
            mutate(tpm = t/sum(t)*1e6) %>%
            ungroup() %>%
  inner_join(eggnogs, by = "orf") %>%
  separate_rows(KEGG_Module, sep = ",") %>%
  group_by(KEGG_Module) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  filter(KEGG_Module != "-") %>%
  arrange(desc(tpm))
```

```{r eggnog_cats_per_treatment, eval=FALSE}
p <- annotated_counts %>%
  separate_rows(best_og_cat, sep = '(?<=.)(?=.)') %>%
  group_by(treatment,best_og_cat) %>%
  summarise(tpm = sum(tpm)) %>%
  ungroup() %>%
  ggplot(mapping = aes(x = treatment, y = tpm, fill = best_og_cat)) +
  geom_col(position = "dodge")

ggplotly(p)

#Categories high in C - P
 # Categories high in low temp - C, J

# Legend
  # P - Inorganic ion transport and metabolism (Maybe membrane potential?)
  # C - Energy production and conversion 
  # J - Translation, ribosomal structure and biogenesis

```

# The gene list and genes expressed
```{r}
bbmap_p %>%
  inner_join(eggnogs, by = "orf") %>%
  inner_join(target_genes %>% rename(Preferred_name = "gene"), by = "Preferred_name") %>%
  inner_join(sample_ID, by = "sample") %>%
  group_by(timepoint, treatment, Preferred_name) %>%
  summarise(tpm = mean(tpm)) %>%
  ungroup() %>%
  inner_join(target_genes %>% rename(Preferred_name = "gene")) %>%
  ggplot(mapping = aes(x = Preferred_name, y = tpm, group = timepoint, fill = timepoint)) +
  geom_col(position = "dodge") +
  theme(axis.text.x = element_text(angle = 60, hjust = 1)) +
  facet_wrap(~ treatment)
```

#Investigating the differentially expressed genes
```{r}
DEGs <- read_tsv("../results/DEG.tsv")
```
## C-treatment
```{r}
C_up <- DEGs %>%
  filter(logFC >= DIFF_TRESHOLD) %>%
  select(orf) %>%
  left_join(eggnogs, by = "orf")

 DEGs %>%
  filter(logFC >= DIFF_TRESHOLD) %>%
  select(orf) %>%
  left_join(eggnogs, by = "orf") %>%
  filter(best_og_desc == "Ammonium Transporter") 

#Among the genes upregulated in the C-treatment, were several ammonium transporters (>10) all belonging to Thaumarchaeota. Different secretion systems, and genes that are part of the lux-regulon (quorom-sensing).
```

## TN-treatment
```{r}
DEGs %>%
  filter(logFC <= -DIFF_TRESHOLD) %>%
  select(orf) %>%
  left_join(eggnogs, by = "orf") %>%
  filter(best_og_desc == "Ammonium transporter") 

#Interestingly, 8 of the downregulated genes (= elevated in the TN treatment) were also ammonium transporters. However these were instead expressed by bacteria. genes involved in growth on fatty acid substrate also found, not present in C-treatment

TN_up <- DEGs %>%
  filter(logFC <= -DIFF_TRESHOLD) %>%
  select(orf) %>%
  left_join(eggnogs, by = "orf")
```

# Results
